ghcide/src/Development/IDE/Plugin/TypeLenses.hs

```diff
@@ -294,3 +324,116 @@ pprPatSynTypeWithoutForalls p = pprPatSynType pWithoutTypeVariables
+        findWhere = grhssLocalBinds
+
+-- | Find all bindings for **one** where clasure.
+findBindingsQ :: GenericQ (Maybe WhereBindings)
+findBindingsQ = something (mkQ Nothing findBindings)
+    where
+        findBindings :: NHsValBindsLR GhcTc -> Maybe WhereBindings
+        findBindings (NValBinds binds sigs) =
+            Just $ WhereBindings
+                { bindings = mapMaybe (something (mkQ Nothing findBindingIds) . snd) binds
+                , existedSigNames = concatMap findSigIds sigs
+                }
+
+        findBindingIds :: LHsBindLR GhcTc GhcTc -> Maybe WhereBinding
+        findBindingIds (L l FunBind{..}) = Just $ WhereBinding (pure $ unLoc fun_id) l
+        findBindingIds (L l PatBind{..}) =
+            let ids = (everything (<>) $ mkQ [] (maybeToList . findIdFromPat)) pat_lhs
+            in Just $ WhereBinding ids l
+        findBindingIds _ = Nothing
+
+        -- | Example: Find `a` and `b` from @(a,b) = (1,True)@
+        findIdFromPat :: Pat GhcTc -> Maybe Id
+        findIdFromPat (VarPat _ (L _ id)) = Just id
+        findIdFromPat _                   = Nothing
+
+        findSigIds (L _ (TypeSig _ names _)) = map unLoc names
+        findSigIds _                         = []
+
+-- | Provide code lens for where bindings.
+whereClauseCodeLens :: PluginMethodHandler IdeState TextDocumentCodeLens
+whereClauseCodeLens state plId CodeLensParams{..} = do
+    enabled <- usePropertyLsp #whereLensOn plId properties
+    if not enabled then pure $ pure $ List [] else pluginResponse $ do
+        nfp <- getNormalizedFilePath plId uri
+        tmr <- handleMaybeM "Unable to typechecking"
+            $ liftIO
+            $ runAction "codeLens.local.TypeCheck" state
+            $ use TypeCheck nfp
+        (hscEnv -> hsc) <- handleMaybeM "Unable to get GhcSession"
+            $ liftIO
+            $ runAction "codeLens.local.GhcSession" state
+            $ use GhcSession nfp
+        let tcGblEnv = tmrTypechecked tmr
+            rdrEnv = tcg_rdr_env tcGblEnv
+            typeCheckedSource = tcg_binds tcGblEnv
+
+            wheres = findWhereQ typeCheckedSource
+            localBindings = mapMaybe findBindingsQ wheres
+
+            -- | Note there may multi ids for one binding
+            bindingToLenses ids span = case srcSpanToRange span of
+                Nothing -> pure []
+                Just range -> forM ids $ \id -> do
+                    (_, fromMaybe [] -> sig) <- liftIO
+                        $ initTcWithGbl hsc tcGblEnv ghostSpan
+                        $ bindToSig id hsc rdrEnv
+                    pure $ generateWhereLens plId range (T.pack sig)
+
+        lenses <- concat <$> sequence
+            [ bindingToLenses idsWithoutSig bindingLoc
+            | WhereBindings{..} <- localBindings
+            , let sigSpans = getSrcSpan <$> existedSigNames
+            , WhereBinding{..} <- bindings
+            , let idsWithoutSig = filter (\x -> getSrcSpan (idName x) `notElem` sigSpans) bindingId
```

Comment:
this seems a bit odd to me: we're checking that a binding doesn't have a signature by comparing source spans. Why not just compare the ids directly?

Reply:
That seems surprising? I would really have expected that the _names_ of the identifiers would be the same? Or maybe there's a different comparison method you're supposed to use there?